core/src/main/scala/caliban/execution/Executor.scala

```diff
@@ -58,11 +58,11 @@ object Executor {
         case s @ PureStep(value) =>
           value match {
             case EnumValue(v) if mergeFields(currentField, v).collectFirst {
-                  case Field("__typename", _, _, _, _, _, _) => true
+                  case Field("__typename", _, _, _, _, _, _, _) => true
                 }.nonEmpty =>
               // special case of an hybrid union containing case objects, those should return an object instead of a string
               val obj = mergeFields(currentField, v).collectFirst {
-                case Field(name @ "__typename", _, _, alias, _, _, _) =>
+                case Field(name @ "__typename", _, _, alias, _, _, _, _) =>
                   ObjectValue(List(alias.getOrElse(name) -> StringValue(v)))
               }
               obj.fold(s)(PureStep(_))
```

Comment:
In fact I am proposing this simplification for current lines 60-68, which pass the tests.
```
        case s @ PureStep(value) =>
          value match {
            case EnumValue(v) =>
              // special case of an hybrid union containing case objects, those should return an object instead of a string
              val obj = mergeFields(currentField, v).collectFirst {
                case f: Field if f.name == "__typename" =>
                  ObjectValue(List(f.alias.getOrElse(f.name) -> StringValue(v)))
              }
              obj.fold(s)(PureStep(_))
            case _ => s
          }
```
I don't see any need to do `mergeFields...collectFirst` twice here, once in match guard and once in evaluation optional `obj`.

Reply:
Note that I am considering `s` to be the default value for `value match` in the complex case and the second match all case just like existing code.